src/Utilities/Compiler/Options/SymbolNamesWithValueOption.cs

```diff
@@ -43,89 +81,287 @@ private SymbolNamesWithValueOption()
+        public bool TryGetValue(ISymbol symbol, [NotNullWhen(true)] out TValue value)
+        {
+            if (_symbols.TryGetValue(symbol, out value) || _names.TryGetValue(symbol.Name, out value))
+            {
+                return true;
+            }
+
+            if (TryGetFirstWildcardMatch(symbol, out var match))
+            {
+                value = match.Value;
+                return true;
+            }
+
+#pragma warning disable CS8653 // A default expression introduces a null value for a type parameter.
+            value = default;
+#pragma warning restore CS8653 // A default expression introduces a null value for a type parameter.
+            return false;
+        }
 
         public override bool Equals(object obj) => Equals(obj as SymbolNamesWithValueOption<TValue>);
 
         public bool Equals(SymbolNamesWithValueOption<TValue>? other)
-            => other != null && _names.IsEqualTo(other._names) && _symbols.IsEqualTo(other._symbols);
+            => other != null && _names.IsEqualTo(other._names) && _symbols.IsEqualTo(other._symbols) && _wildcardNamesBySymbolKind.IsEqualTo(other._wildcardNamesBySymbolKind);
 
         public override int GetHashCode()
-            => HashUtilities.Combine(HashUtilities.Combine(_names), HashUtilities.Combine(_symbols));
+            => HashUtilities.Combine(HashUtilities.Combine(_names), HashUtilities.Combine(_symbols), HashUtilities.Combine(_wildcardNamesBySymbolKind));
+
+        private bool TryGetFirstWildcardMatch(ISymbol symbol, out KeyValuePair<string, TValue> firstMatch)
+        {
+            firstMatch = NoWildcardMatch;
+
+            if (_wildcardNamesBySymbolKind.IsEmpty)
+            {
+                return false;
+            }
+
+            Debug.Assert(symbol.Kind == SymbolKind.Event || symbol.Kind == SymbolKind.Field || symbol.Kind == SymbolKind.Method || symbol.Kind == SymbolKind.NamedType || symbol.Kind == SymbolKind.Namespace || symbol.Kind == SymbolKind.Property);
+
+            // The matching was already processed
+            if (_wildcardMatchResult.ContainsKey(symbol))
+            {
+                firstMatch = _wildcardMatchResult[symbol];
+                return !firstMatch.Equals(NoWildcardMatch);
+            }
+
+            var symbolFullNameBuilder = new StringBuilder();
+            var symbolKindsToCheck = new HashSet<SymbolKind> { symbol.Kind };
+
+            // Try a partial check on the symbol...
+            if (TryGetSymbolPartialMatch(symbolFullNameBuilder, symbol, out firstMatch))
+            {
+                return true;
+            }
+
+            // ...Now try a partial check looping on the symbol's containing type...
+            INamedTypeSymbol? currentType = symbol.ContainingType;
+            while (currentType != null)
+            {
+                if (TryGetSymbolPartialMatch(symbolFullNameBuilder, currentType, out firstMatch))
```

Comment:
I would have taken a much more simpler implementation approach. The cache is defined with `INamespaceOrTypeSymbol` as the key. In both the while loops here that walk the containing type and namespace chain, use `_wildcardMatchResult.GetOrAdd` to get this bool. The func passed for Add would invoke TryGetSymbolPartialMatch. `_wildcardMatchResult` does not need to be used at any other place except inside these two loops.

Additionally, also consider making the local functions static in this method to avoid accidental capture or update. You can probably pass `_wildcardNamesBySymbolKind` as an explicit argument to local function invocations.